src/IB/CIBStochasticMethod.cpp

```diff
@@ -0,0 +1,463 @@
+                            0.5 * dt * d_trans_vel_half[struct_handle][d];
+
+                if (periodic_shift[d])
+                {
+                    while (X_half[d] < domain_x_lower[d])
+                    {
+                        X_half[d] += domain_length[d];
+                    }
+                    while (X_half[d] >= domain_x_upper[d])
+                    {
+                        X_half[d] -= domain_length[d];
+                    }
+                }
+            }
+        }
+        (*X_half_data)[ln]->restoreArrays();
+        d_l_data_manager->getLData("X0_unshifted", ln)->restoreArrays();
+    }
+    *X_half_needs_ghost_fill = true;
+
+    // Compute the COM at mid-step.
+    for (unsigned struct_no = 0; struct_no < d_num_rigid_parts; ++struct_no)
+    {
+        for (unsigned int d = 0; d < NDIM; ++d)
+        {
+            d_center_of_mass_half[struct_no][d] =
+                d_center_of_mass_current[struct_no][d] + 0.5 * dt * d_trans_vel_half[struct_no][d];
+            if (periodic_shift[d])
+            {
+                while (d_center_of_mass_half[struct_no][d] < domain_x_lower[d])
+                {
+                    d_center_of_mass_half[struct_no][d] += domain_length[d];
+                }
+                while (d_center_of_mass_half[struct_no][d] >= domain_x_upper[d])
+                {
+                    d_center_of_mass_half[struct_no][d] -= domain_length[d];
+                }
+            }
+        }
+    }
+    return;
+} // moveLagrangianData
+
+void
+CIBStochasticMethod::computeRFDforcesAndDisplacements()
+{
+    Vec F_rfd, U_rfd;
+    double Vel_scale = (1.0 / d_kT);
+    getNetExternalForceTorque(&F_rfd, d_current_time);
+    VecDuplicate(F_rfd, &U_rfd);
+    VecCopy(F_rfd, U_rfd);
+    VecScale(U_rfd, Vel_scale);
+
+    Vec U_all;
+    VecScatter ctx;
+    VecScatterCreateToAll(U_rfd, &ctx, &U_all);
+    VecScatterBegin(ctx, U_rfd, U_all, INSERT_VALUES, SCATTER_FORWARD);
+    VecScatterEnd(ctx, U_rfd, U_all, INSERT_VALUES, SCATTER_FORWARD);
+    const PetscScalar* U_array;
+    VecGetArrayRead(U_all, &U_array);
+
+    int part_free_dofs_begin = 0;
+    for (unsigned part = 0; part < d_num_rigid_parts; ++part)
+    {
+        int num_free_dofs;
+        const FreeRigidDOFVector& solve_dofs = getSolveRigidBodyVelocity(part, num_free_dofs);
+        if (!num_free_dofs) continue;
+
+        RDV UW;
+        eigenToRDV(d_trans_vel_current[part], d_rot_vel_current[part], UW);
+        const PetscScalar* a = &U_array[part_free_dofs_begin];
+        for (int k = 0, p = 0; k < s_max_free_dofs; ++k)
+        {
+            if (solve_dofs[k])
+            {
+                UW[k] = a[p];
+                ++p;
+            }
+        }
+        rdvToEigen(UW, d_trans_vel_current[part], d_rot_vel_current[part]);
+        d_trans_vel_current[part] *= (d_L_scale * d_L_scale);
+        part_free_dofs_begin += num_free_dofs;
+    }
```

Comment:
I don't fully understand what this does but this algorithm seems suspicious in that every single processor starts at the beginning of the global array and then does the same loop. Is there any way to avoid `U_all` here?

Reply:
@brennansprinkle: What is the intent here? The free rigid degrees of freedom (that can actually change from time step to time step --- motion can be constrained or become free at any tilmestep) vector is a global vector with all the DoFs owned by processor zero as this is a very small vector. For a single body in the domain, the size is 3 in 2D and 6 in 3D. This is how I set it

https://github.com/IBAMR/IBAMR/blob/master/src/IB/CIBMethod.cpp#L263-L305

Reply:
@drwells: We create these MPI vectors and turn it to PETSc to do iterative solves.

Reply:
Sorry, this function is entirely vestigial and should be removed as I don't believe that it is called anywhere. The original functionality (requiring a helper function that I did remove) was replaced by CIBStochasticStokesSolver::computeFUforRFD.

Reply:
@brennansprinkle : Is there a better name than `FU`. Reminds me of Frank Underwood and the obvious.

Reply:
Yeah, if I'm being honest that did cross my mind, but seemed more of a feature than a bug to me. Still, you're probably right :-) Maybe 'computeRFDForcesAndVelocities'

Reply:
What is RFD? Rigid Fluctuating Dynamics?

Reply:
Random Finite Difference.
Sorry, yeah that's a bit jargony but it's what we call it in the paper so I think it should stay in there

Reply:
Agree. We should put RFD in the class description.